lib/src/main/kotlin/tree/AVLTree.kt

```diff
@@ -2,16 +2,142 @@ package tree
+
+        return nodeRight
+    }
+
+    private fun rebalanced(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>?): AVLTreeNode<K, V>?{
+        updateHeight(root)
+        val balance = root?.let { getBalance(it) }
+        val rootLeft = root?.left
+        val rootRight = root?.right
+
+        if (node != null && balance != null){
+            if (balance > 1) {
+                if (rootLeft != null) {
+                    if (node.key < rootLeft.key) {
+                        return rotateRight(root)
+                    } else {
+                        root.left = root.left?.let { rotateLeft(it) }
+                        return rotateRight(root)
+                    }
+                }
+            }
+            if (balance < -1) {
+                if (rootRight != null) {
+                    if (node.key > rootRight.key) {
+                        return rotateLeft(root)
+                    } else {
+                        root.right = root.right?.let { rotateRight(it) }
+                        return rotateLeft(root)
+                    }
+                }
+            }
+        }
+
+        return root
+    }
+
+    private fun insertRecursive(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>?): AVLTreeNode<K, V>? {
+        if (root == null) {
+            return node
+        }
+
+        if (node != null) {
+            if (root.key > node.key) {
+                root.left = insertRecursive(root.left, node)
+            } else  if (root.key < node.key) {
+                    root.right = insertRecursive(root.right, node)
+            }
+        }
+
+        return rebalanced(root, node)
+    }
+
     override fun insertNode(node: AVLTreeNode<K, V>) {
-        TODO("Insert node to tree")
+        root = insertRecursive(root, node)
+    }
+
+    private fun minValueNode(node: AVLTreeNode<K, V>): AVLTreeNode<K, V> {
+        var current = node
+        val curLeft = current.left
+        while (curLeft != null) current = curLeft
+        return current
+    }
+    private fun removeRecursive(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        var rootNode = root
+
+        if (rootNode == null) {
+            return root
+        }
+
+        if (rootNode.key > node.key) {
+            rootNode.left = removeRecursive(rootNode.left, node)
+        } else if (rootNode.key < node.key) {
+            rootNode.right = removeRecursive(rootNode.right, node)
+        } else {
+            val rootNodeRight = rootNode.right
+            val rootNodeLeft = rootNode.left
+            if (rootNode.left == null || rootNode.right == null) {
+                rootNode = if (rootNode.left == null) rootNodeRight else rootNodeLeft
+            } else {
+                val mostLeftChild: AVLTreeNode<K, V>? = rootNode.right?.let { minValueNode(it) }
+                rootNode = mostLeftChild
+                rootNode?.right = rootNode?.let { removeRecursive(rootNode.right, it) }
+            }
+        }
```

Comment:
Предлагаю избавиться от вложенности, выходя из if с помощью return и отделив случай разных ключей у rootNode и node

```suggestion
        if (rootNode.key > node.key) {
            rootNode.left = removeRecursive(rootNode.left, node)
            return rebalanced(rootNode, node)
        } else if (rootNode.key < node.key) {
            rootNode.right = removeRecursive(rootNode.right, node)
            return rebalanced(rootNode, node)
        }

        val rootNodeRight = rootNode.right
        val rootNodeLeft = rootNode.left
        if (rootNode.left == null || rootNode.right == null) {
            rootNode = if (rootNode.left == null) rootNodeRight else rootNodeLeft
        } else {
            val mostLeftChild: AVLTreeNode<K, V>? = rootNode.right?.let { minValueNode(it) }
            rootNode = mostLeftChild
            rootNode?.right = rootNode?.let { removeRecursive(rootNode.right, it) }
        }
```